From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: roberth <roberth@arista.com>
Date: Tue, 6 Feb 2024 17:50:49 -0600
Subject: [PATCH] Sign the image with the development key

This adds the necessary script to automatically sign the image with the
development key. No linux/bash compatible scripts are provided by AMD for this,
but there is a batch script
   AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/BuildPspImageSign.bat
for Windows that was used as a reference.

The private key used is located at
   AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/TestRtmPrivateKey.pem
and should be replaced before executing the build.

This should be enough to get the image to pass PSB test mode.
---
 .../AMDTools/NewPspKit/BuildPspImage.sh       |  5 ++
 .../NewPspKit/arista_gen_rtm_signature.sh     | 56 +++++++++++++++++++
 .../BIOSImageDirectory32M_DDR5.xml            | 14 ++---
 .../BIOSImageDirectory32M_DDR5_L2A.xml        |  2 +-
 Board/AMD/FoxBoardPkg/PspBuild.sh             | 11 +---
 5 files changed, 71 insertions(+), 17 deletions(-)
 create mode 100644 AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/arista_gen_rtm_signature.sh

diff --git a/AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/BuildPspImage.sh b/AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/BuildPspImage.sh
index 14be5b74..82641080 100644
--- a/AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/BuildPspImage.sh
+++ b/AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/BuildPspImage.sh
@@ -55,6 +55,7 @@ InputBiosFile=$1
 RTM_FILE=$2
 PspDirectoryCfg=$3
 FINAL_BIOS=$4
+KEY_MODE=$5
 
 #Related PATH
 PSP_DIR_TOOL_PATH=$PSPKIT_PATH/PspDirectoryTool
@@ -86,6 +87,10 @@ this_command="python3 $BUILDPSPDIRECTORY bd $InputBiosFile $PspDirectoryCfg"
 echo "$this_command"
 eval $this_command || ERROR
 
+if [ "$KEY_MODE" == "TK" ]; then
+   echo "[Generate RTM Signature]"
+   ./arista_gen_rtm_signature.sh "$RTM_FILE" "$RTM_PRIVATE_KEY" sha384
+fi
 
 echo ""
 echo "[Finally build the Final BiosImage]"
diff --git a/AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/arista_gen_rtm_signature.sh b/AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/arista_gen_rtm_signature.sh
new file mode 100644
index 00000000..6e73b00d
--- /dev/null
+++ b/AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/arista_gen_rtm_signature.sh
@@ -0,0 +1,56 @@
+#!/bin/bash
+# Copyright (c) 2024 Arista Networks, Inc.  All rights reserved.
+# Arista Networks, Inc. Confidential and Proprietary.
+
+# Adapted from AMD/AmdEmbedded/AgesaModulePkg/AMDTools/NewPspKit/BuildPspImageSign.bat
+
+set -e
+
+RTM_FILE="$1"
+RTM_PRIVATE_KEY="$2"
+SHATYPE="$3" # sha256 or sha384
+
+if [ "$SHATYPE" = "sha256" ]; then
+   RSA_PSS_SALTLEN=-2
+elif [ "$SHATYPE" = "sha384" ]; then
+   RSA_PSS_SALTLEN=-1
+else
+   echo "Unknown algorithm: $SHATYPE" >&2
+   exit 1
+fi
+
+# These specific file names are generated by AMD's tools
+BIOS_L2_DIR_FILE=Output/BiosDirHeaderL2.bin
+BIOS_L2_DIR_B_FILE=Output/BiosDirHeaderL2B.bin
+
+# Signature for single partition or partition A
+RTM_FILE_SIGNATURE=RTMSignature.bin
+# Signature for partition B (dual boot only)
+RTM_FILE_SIGNATURE_B=RTMSignatureB.bin
+
+sign_with_openssl() {
+   local signature="$1"
+   local data="$2"
+   local reverse_sig="$(mktemp)"
+   openssl dgst -binary -sign "$RTM_PRIVATE_KEY" -"$SHATYPE"                        \
+      -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:"$RSA_PSS_SALTLEN"       \
+      -out "$reverse_sig" "$data"
+   # Signature needs to be stored in little-endian byte order
+   python -c 'import sys; sys.stdout.buffer.write(sys.stdin.buffer.read()[::-1])'   \
+      <"$reverse_sig" >"$signature"
+   rm -f "$reverse_sig"
+}
+
+PAYLOAD="$(mktemp)"
+
+cat "$RTM_FILE" "$BIOS_L2_DIR_FILE" >"$PAYLOAD"
+echo "Signing RTM+BIOS_DIR_L2"
+sign_with_openssl "$RTM_FILE_SIGNATURE" "$PAYLOAD"
+
+if [ -f "$BIOS_L2_DIR_B_FILE" ]; then
+   cat "$RTM_FILE" "$BIOS_L2_DIR_B_FILE" >"$PAYLOAD"
+   echo "Signing RTM+BIOS_DIR_L2B"
+   sign_with_openssl "$RTM_FILE_SIGNATURE_B" "$PAYLOAD"
+fi
+
+rm -f "$PAYLOAD"
diff --git a/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5.xml b/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5.xml
index 589c9c19..f4ca38fa 100644
--- a/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5.xml
+++ b/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5.xml
@@ -32,7 +32,7 @@
   <ISH_HEADER AddressMode="0x1" Base="0x000C4100" PspId="0xbc0d0200" Location="0x000C5000" BootPriority="0xffffffff" GlitchRetries="0x0" Reserved_1="0xffffffff" SlotMaxSize="0xffffffff" UpdateRetries="0xf" SpiBlockSize="0x1000"/>
 
 <!-- Image slot A PSP DIR-->
-  <PSP_DIR Level="2" Base="0x000C5000" Size="0x00380000" SpiBlockSize="0x1000">
+  <PSP_DIR Level="2" LevelType="A" Base="0x000C5000" Size="0x00380000" SpiBlockSize="0x1000">
     <!-- AMD Public Key -->
     <IMAGE_ENTRY Type="0x00" File="TypeId0x00_RootKey_RMB.tkn"/>
     <!-- PSP Boot Loader firmware -->
@@ -140,14 +140,14 @@
   </PSP_DIR>
 
 <!-- Image slot A BIOS DIR-->
-  <BIOS_DIR Level="2" Base="0x00445000" Size="0x00048000" SpiBlockSize="0x1000">
+  <BIOS_DIR Level="2" LevelType="A" Base="0x00445000" Size="0x00048000" SpiBlockSize="0x1000">
     <!-- Refer to BIOS guide for detail requirement of each Type -->
     <!-- All definition is same as PSP_DIR except Destination -->
     <!-- Attribute Destination:(Optional)-->
     <!-- Destination: Used to specify the destination memory address PSP FW need copy to -->
     <IMAGE_ENTRY Type="0x05" File="TypeId0x05_RtmPubSigned_RMB.key"/>
     <!-- The RTMSignature.bin is NOT ready when generate Header with "BD" command, Use explicit "Size" to avoid open file fail  -->
-    <!-- <IMAGE_ENTRY Type="0x07" File="RTMSignature.bin" Size="0x100"/> -->
+    <IMAGE_ENTRY Type="0x07" File="RTMSignature.bin" Size="0x200"/>
 
     <!-- Agesa PSP Customization Block (PSP Agesa variable) NVRAM (APCB)-->
     <IMAGE_ENTRY Type="0x60" Instance="0x00" File="APCB_FP7r2_Updatable.bin" Size="0x3000"/>
@@ -201,7 +201,7 @@
   </BIOS_DIR>
 
 <!-- Image slot B PSP DIR-->
-  <PSP_DIR Level="2" Base="0x0048D000" Size="0x00380000" SpiBlockSize="0x1000">
+  <PSP_DIR Level="2" LevelType="B" Base="0x0048D000" Size="0x00380000" SpiBlockSize="0x1000">
     <!-- AMD Public Key -->
     <IMAGE_ENTRY Type="0x00" File="TypeId0x00_RootKey_RMB.tkn"/>
     <!-- PSP Boot Loader firmware -->
@@ -306,14 +306,14 @@
   </PSP_DIR>
 
 <!-- Image slot B BIOS DIR-->
-  <BIOS_DIR Level="2" Base="0x0080D000" Size="0x00048000" SpiBlockSize="0x1000">
+  <BIOS_DIR Level="2" LevelType="B" Base="0x0080D000" Size="0x00048000" SpiBlockSize="0x1000">
     <!-- Refer to BIOS guide for detail requirement of each Type -->
     <!-- All definition is same as PSP_DIR except Destination -->
     <!-- Attribute Destination:(Optional)-->
     <!-- Destination: Used to specify the destination memory address PSP FW need copy to -->
     <IMAGE_ENTRY Type="0x05" File="TypeId0x05_RtmPubSigned_RMB.key"/>
     <!-- The RTMSignature.bin is NOT ready when generate Header with "BD" command, Use explicit "Size" to avoid open file fail  -->
-    <!-- <IMAGE_ENTRY Type="0x07" File="RTMSignature.bin" Size="0x100"/> -->
+    <IMAGE_ENTRY Type="0x07" File="RTMSignatureB.bin" Size="0x200"/>
 
     <!-- Agesa PSP Customization Block (PSP Agesa variable) NVRAM (APCB)-->
     <IMAGE_ENTRY Type="0x60" Instance="0x00" File="APCB_FP7r2_Updatable.bin" Size="0x3000"/>
@@ -365,4 +365,4 @@
     <POINT_ENTRY Type="0x6D" Instance="4" Address="0x01F8E000" Size="0x00001000" AddressMode="1"/>
 
   </BIOS_DIR>
-</DIRS>
\ No newline at end of file
+</DIRS>
diff --git a/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5_L2A.xml b/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5_L2A.xml
index 24736fc0..deb5107b 100644
--- a/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5_L2A.xml
+++ b/Board/AMD/FoxBoardPkg/PlatformConfig/BIOSImageDirectory32M_DDR5_L2A.xml
@@ -131,7 +131,7 @@
     <!-- Destination: Used to specify the destination memory address PSP FW need copy to -->
     <IMAGE_ENTRY Type="0x05" File="TypeId0x05_RtmPubSigned_RMB.key"/>
     <!-- The RTMSignature.bin is NOT ready when generate Header with "BD" command, Use explicit "Size" to avoid open file fail  -->
-    <!-- <IMAGE_ENTRY Type="0x07" File="RTMSignature.bin" Size="0x100"/> -->
+    <IMAGE_ENTRY Type="0x07" File="RTMSignature.bin" Size="0x200"/>
 
     <!-- Agesa PSP Customization Block (PSP Agesa variable) NVRAM (APCB)-->
     <IMAGE_ENTRY Type="0x60" Instance="0x00" File="APCB_FP7r2_Updatable.bin" Size="0x3000"/>
diff --git a/Board/AMD/FoxBoardPkg/PspBuild.sh b/Board/AMD/FoxBoardPkg/PspBuild.sh
index 5fa742c2..6f5a179b 100755
--- a/Board/AMD/FoxBoardPkg/PspBuild.sh
+++ b/Board/AMD/FoxBoardPkg/PspBuild.sh
@@ -13,15 +13,8 @@ echo KEY_MODE $KEY_MODE
 PSP_PLATFORM_PATH=$WORKSPACE/$PROJECT_REL_PATH/$PROJECT_PKG/PlatformConfig
 KVM_PLATFORM_PATH=$WORKSPACE/$PROJECT_REL_PATH/$PROJECT_PKG/Binary/KVM
 
-if [ $KEY_MODE = "PK" ]; then
-  PSP_FW_PATH=$WORKSPACE/$CHIPSET_REL_PATH/AgesaModulePkg/Firmwares/RMB
-  PSP_CONFIG_FILE_PATH=$PSP_PLATFORM_PATH
-fi
-
-if [ $KEY_MODE = "NOSIGN" ]; then
-  PSP_FW_PATH=$WORKSPACE/$CHIPSET_REL_PATH/AgesaModulePkg/Firmwares/RMB
-  PSP_CONFIG_FILE_PATH=$PSP_PLATFORM_PATH
-fi
+PSP_FW_PATH=$WORKSPACE/$CHIPSET_REL_PATH/AgesaModulePkg/Firmwares/RMB
+PSP_CONFIG_FILE_PATH=$PSP_PLATFORM_PATH
 
 echo PSP_FW_PATH $PSP_FW_PATH
 
-- 
2.41.0

